test/modules/packages/Toml/Toml.chpl

```diff
@@ -0,0 +1,757 @@
+         }
+         }
+     }
+
+
+     /* 
+      For the user to write values of a node as follows:
+      Node[key].toString()
+     */
+      proc toString() : string { 
+        return toString(this);
+     }
+
+     
+     proc deinit() {
+       for a in A {
+         delete a;
+       }
+     }
+   } 
+ }
+
+
+
+/*
+Reader module for use in the Parser Class. 
+*/
+ module TomlReader {
+ 
+   
+   /* Returns the next token in the current line without removing it */
+   proc top(source) {
+     if source.currentLine.D.size < 1 {
+       source.newLine();
+     }
+     return source.currentLine[source.currentLine.D.first];
+   }
+   
+   /* Returns a boolean or wether or not another line can be read
+   /  Also updates the currentLine if empty */
+   proc readLine(source) {
+     return source.nextLine();
+   }
+   
+   proc skipNext(source) { 
+     source.currentLine.skip();
+   }
+   
+   proc skipLine(source) {
+     source.skipROL();
+   }
+   
+   /* retrives the next token in currentline */
+   proc getToken(source) {
+     return source.nextToke();
+   }
+   
+
+   class Source {
+     
+     var tomlFile: string;
+     var tokenD = {1..0},
+       tokenlist: [tokenD] Tokens;
+     var currentLine: Tokens;
+     
+     
+     proc init(tomlFile: string) {
+       this.tomlFile = tomlFile;
+       var openFile = open(tomlFile, iomode.r);
+       this.genTokenlist(openFile);
+     }
+
+
+     /* 
+     proc init(tomlStr: string) {
+      this.tomlStr = tomlStr;
+      this.genTokenlist(tomlStr);
+     }
+      */
+     
+     /* generates list of Token objects */
+     proc genTokenlist(input) {
+       for line in input.lines() {
+         splitLine(line);
+       }
+       currentLine = tokenlist[tokenD.first];
+     }
+     
+     /*
+     proc genTokenlist(input: string) {
+       for line in input.split('\n') {
+         splitLine(line);
+       }
+      currentLine = tokenlist[tokenD.first];
+     } 
+     */
+  
+     proc splitLine(line) {
+       var linetokens: [1..0] string;
+       const doubleQuotes = '".*?\\[^\\]?"',
+          singleQuotes = "'.*?\\[^\\]?'",
+          bracketContents = "(\\[\\w+\\])",
+          brackets = "(\\[)|(\\])",
+          comments = "(\\#)",
+          commas = "(\\,)",
+          equals = "(\\=)",
+          curly = "(\\{)|(\\})";
+
+       const pattern = compile('|'.join(doubleQuotes,
+				     singleQuotes,
+				     bracketContents,
+				     brackets,
+				     comments,
+				     commas,
+                                     curly,
+				     equals));
+
+       for token in pattern.split(line) {
+         var strippedToken = token.strip();
+         if strippedToken.length != 0  {
+           linetokens.push_back(strippedToken);
+         }
+       }
+       if linetokens.size != 0 {
+         var tokens = new Tokens(linetokens);
+         tokenlist.push_back(tokens);
+       }
+     }
+     
+     
+     proc newLine() {
+       if nextLine() { 
+         if currentLine.D.size < 1 {
```

Comment:
might be worth creating a helper method for this: `proc Tokens.isEmpty()` and replacing the `currentLine.D.size < 1` and `currentLine.A.isEmpty()` calls with it.